### PR TITLE
feat(rest): New search resource endpoint and get releases for multiple projects

### PR DESCRIPTION
**Pull request by smrutis1**
[//]: # (Copyright Bosch.IO GmbH 2020)
[//]: # (This program and the accompanying materials are made)
[//]: # (available under the terms of the Eclipse Public License 2.0)
[//]: # (which is available at https://www.eclipse.org/legal/epl-2.0/)
[//]: # (SPDX-License-Identifier: EPL-2.0)

> Please provide a summary of your changes here.

Added two new endpoints viz. to search resources using `api/search?searchText=376570&typeMasks=project` and to get releases using multiple project ids. `/projects/releases?clearingState=APPROVED` and added pagination support to the later one along with the get all releases endpoint.

Issue: 

Need to test newly created rest endpoints as mentioned in the documentation.

> Have you implemented any additional tests? 

 Yes, for REST documentation

### Checklist
Must:
- [ ] All related issues are referenced in commit messages and in PR

Signed-off-by: Smruti Prakash Sahoo <smruti.sahoo@siemens.com>

### CODE REVIEW

**Review by JaideepPalit — CHANGES_REQUESTED** [review]
Reviewed the PR - Kindly look into some review comments.

File: rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/core/RestControllerHelper.java [hunk 1]
```diff
@@ -437,6 +441,31 @@ public Release convertToEmbeddedRelease(Release release) {
         return embeddedRelease;
     }
 
+    public Release convertToEmbeddedReleaseWithDet(Release release) {
+        Release embeddedRelease = new Release();
+        embeddedRelease.setId(release.getId());
```
[the remark this answers is not in the document]
Comment (smrutis1) [reply rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/core/RestControllerHelper.java]: reused

File: rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/search/SearchController.java [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+    
+    @Autowired
+    private Sw360SearchService sw360SearchService;
+    
+    @NonNull
+    private final RestControllerHelper restControllerHelper;
+    
+
+    @RequestMapping(value = SEARCH_URL, method = RequestMethod.GET)
+    public ResponseEntity<Resources<Resource<SearchResult>>> getSearchResult(@RequestParam(value = "searchText") String searchText, @RequestParam Optional<List<String>> typeMasks) throws TException {
```
[the remark this answers is not in the document]
Comment (smrutis1) [reply rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/search/SearchController.java]: added

File: rest/resource-server/src/test/java/org/eclipse/sw360/rest/resourceserver/restdocs/ProjectSpecTest.java [hunk 3]
```diff
@@ -1111,6 +1140,64 @@ public void should_document_get_download_license_info() throws Exception {
                                 parameterWithName("externalIds").description("The external Ids of the project"))));
     }
 
+    @Test
+    public void should_document_get_projects_releases() throws Exception {
+        String accessToken = TestHelper.getAccessToken(mockMvc, testUserId, testUserPassword);
+        this.mockMvc.perform(get("/api/projects/releases")
+                 .header("Authorization", "Bearer " + accessToken)
+                 .param("clearingState", ClearingState.APPROVED.toString())
+                 .param("transitive", "false")
```
[the remark this answers is not in the document]
Comment (smrutis1) [reply rest/resource-server/src/test/java/org/eclipse/sw360/rest/resourceserver/restdocs/ProjectSpecTest.java:1149]: Added the request statement

File: rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/project/Sw360ProjectService.java [hunk 4]
```diff
@@ -270,4 +278,69 @@ public Project convertToEmbeddedWithExternalIds(Project sw360Object) {
...
+                List<Release> releases = releaseIds.stream().map(relId -> wrapTException(() -> {
+                    final Release sw360Release = releaseService.getReleaseForUserById(relId, sw360User);
+                    return sw360Release;
+                })).collect(Collectors.toList());
+                return releases;
+            };
+            callableTasksToGetReleases.add(getReleasesByProjectId);
+        });
+
+        ExecutorService  exeService = Executors.newCachedThreadPool();
```
[the remark this answers is not in the document]
Comment (smrutis1) [reply rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/project/Sw360ProjectService.java]: Have used fixed thread pool and made it static

**Comment by smrutis1** [comment]
@JaideepPalit , Thanks for review. Have incorporated the review comments.

**Review by JaideepPalit — CHANGES_REQUESTED** [review]
Tested the PR. Some Observations.
- Search with project , component, release , license, vendor - `working fine `
- Get project releases based on clearing state - `working fine`


- Search with user and obligation is not working, always 204 No content is coming.
- Exception is coming at tomcat startup
```
2020-11-05 06:23:40.249 ERROR [Framework Event Dispatcher: Equinox Container: 0dd3384e-20bb-4cdb-ad33-690b20cf17de][Framework:93] FrameworkEvent ERROR 
java.util.ConcurrentModificationException
	at java.base/java.util.HashMap.computeIfAbsent(HashMap.java:1134)
	at org.apache.aries.component.dsl.OSGi.lambda$null$79(OSGi.java:734)
	at org.apache.aries.component.dsl.internal.JustOSGiImpl.lambda$new$2(JustOSGiImpl.java:47)
	at org.apache.aries.component.dsl.internal.OSGiImpl.run(OSGiImpl.java:50)
	at org.apache.aries.component.dsl.OSGi.lambda$null$80(OSGi.java:732)
	at org.apache.aries.component.dsl.OSGi.lambda$null$64(OSGi.java:602)
	at org.apache.aries.component.dsl.Publisher.apply(Publisher.java:28)
	at org.apache.aries.component.dsl.OSGi.lambda$null$66(OSGi.java:663)
	at org.apache.aries.component.dsl.Publisher.apply(Publisher.java:28)
```

- Exception is coming at tomcat shutdown.

```
Exception in thread "pool-147-thread-1" java.lang.NoClassDefFoundError: org/apache/logging/log4j/message/ParameterizedMessage
	at org.apache.logging.log4j.message.ParameterizedNoReferenceMessageFactory.newMessage(ParameterizedNoReferenceMessageFactory.java:104)
	at org.apache.logging.log4j.message.AbstractMessageFactory.newMessage(AbstractMessageFactory.java:75)
	at org.apache.logging.log4j.spi.AbstractLogger.logMessage(AbstractLogger.java:2010)
	at org.apache.logging.log4j.spi.AbstractLogger.logIfEnabled(AbstractLogger.java:1884)
	at org.apache.logging.log4j.spi.AbstractLogger.error(AbstractLogger.java:793)
	at org.apache.logging.log4j.core.util.DefaultShutdownCallbackRegistry.run(DefaultShutdownCallbackRegistry.java:76)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: java.lang.ClassNotFoundException: Illegal access: this web application instance has been stopped already. Could not load [org.apache.logging.log4j.message.ParameterizedMessage]. The following stack trace is thrown for debugging purposes as well as to attempt to terminate the thread which caused the illegal access.
	at org.apache.catalina.loader.WebappClassLoaderBase.checkStateForClassLoading(WebappClassLoaderBase.java:1375)
	at org.apache.catalina.loader.WebappClassLoaderBase.loadClass(WebappClassLoaderBase.java:1226)
	at org.apache.catalina.loader.WebappClassLoaderBase.loadClass(WebappClassLoaderBase.java:1188)
	... 7 more
Caused by: java.lang.IllegalStateException: Illegal access: this web application instance has been stopped already. Could not load [org.apache.logging.log4j.message.ParameterizedMessage]. The following stack trace is thrown for debugging purposes as well as to attempt to terminate the thread which caused the illegal access.
	at org.apache.catalina.loader.WebappClassLoaderBase.checkStateForResourceLoading(WebappClassLoaderBase.java:1385)
	at org.apache.catalina.loader.WebappClassLoaderBase.checkStateForClassLoading(WebappClassLoaderBase.java:1373)
	... 9 more
Exception in thread "pool-148-thread-1" Exception in thread "pool-142-thread-1" Exception in thread "pool-144-thread-1" java.lang.NoClassDefFoundError: org/apache/logging/log4j/message/ParameterizedMessage
	at org.apache.logging.log4j.message.ParameterizedNoReferenceMessageFactory.newMessage(ParameterizedNoReferenceMessageFactory.java:104)
	at org.apache.logging.log4j.message.AbstractMessageFactory.newMessage(AbstractMessageFactory.java:75)
	at org.apache.logging.log4j.spi.AbstractLogger.logMessage(AbstractLogger.java:2010)
	at org.apache.logging.log4j.spi.AbstractLogger.logIfEnabled(AbstractLogger.java:1884)
	at org.apache.logging.log4j.spi.AbstractLogger.error(AbstractLogger.java:793)
	at org.apache.logging.log4j.core.util.DefaultShutdownCallbackRegistry.run(DefaultShutdownCallbackRegistry.java:76)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: java.lang.ClassNotFoundException: Illegal access: this web application instance has been stopped already. Could not load [org.apache.logging.log4j.message.ParameterizedMessage]. The following stack trace is thrown for debugging purposes as well as to attempt to terminate the thread which caused the illegal access.
	at org.apache.catalina.loader.WebappClassLoaderBase.checkStateForClassLoading(WebappClassLoaderBase.java:1375)
	at org.apache.catalina.loader.WebappClassLoaderBase.loadClass(WebappClassLoaderBase.java:1226)
	at org.apache.catalina.loader.WebappClassLoaderBase.loadClass(WebappClassLoaderBase.java:1188)
	... 7 more

```

File: rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/core/RestControllerHelper.java [hunk 5]
```diff
@@ -437,6 +439,28 @@ public Release convertToEmbeddedRelease(Release release) {
         return embeddedRelease;
     }
 
+    public Release convertToEmbeddedReleaseWithDet(Release release) {
```
[the remark this answers is not in the document]
Comment (smrutis1) [reply rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/core/RestControllerHelper.java:442]: Done

**Comment by smrutis1** [comment]
> * Search with user and obligation is not working, always 204 No content is coming.

May be need to test it where the user and obligation search is working in UI as there is no change in rest except calling the backend service and giving the response back.

> * Exception is coming at tomcat startup
>

Unfortunately I am not able to reproduce the issue. 

> 
> * Exception is coming at tomcat shutdown.
> 


I have been seeing this for other branches as well. May be not due to the changes in this PR

**Comment by JaideepPalit** [comment]
Deployed on test instance and Retested the PR.

> * Exception is coming at tomcat startup
>    - Exception is not appearing in test instance

> * Exception is coming at tomcat shutdown
>    - Similar exception is coming in separate branch in test instance. Could be related to some other issue.

> * Search with user and obligation is not working, always 204 No content is coming.
>   - Search with user is `working` in test instance.
>   - Search with obligation is `not working` in test instance.

**Comment by JaideepPalit** [comment]
(from discussion) - Created separate issue #1035 for search with Obligation

**Review by JaideepPalit — APPROVED** [review]

**Pull request merged**